KAFKA-2431: Easier Testing of SSL

benstopford opened the pull request:
- Allow cipher suites to be specified relevant properties
- Avoid System.exit in ProducerPerformance so this can be externally invoked
- Add command line option so that a default properties can be specified in ConsumerPerformance (needed for ssl properties)
- Allow cipher suites to be specified via properties
- Increase timeout in ConsumerPerformance (useful when running with SSL under load)
- Use a relative default value for the reporting interval in Consumer Performance
- Add option for show-all-stats (includes intermediary and summary stats) in Consumer Performance

benstopford committed:
KAFKA-2431: Ease testing of SSL

- Avoid System.exit in ProducerPerformance so this can be externally invoked
- Add command line option so that a default set of properties can be speicfied in ConsumerPerformance (needed for ssl properties)
- Allow cipher suites to be specified via properties
- Increase timeout in ConsumerPerformance (useful when running with SSL under load)
- Use a relative default value for the reporting interval
- Add option for show-all-stats (includes intermediary and summary stats)

eribeiro commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SSLFactory.java:
```diff
@@ -60,12 +60,13 @@ public void configure(Map<String, ?> configs) throws KafkaException {
         if (configs.get(SSLConfigs.SSL_CIPHER_SUITES_CONFIG) != null) {
             List<String> cipherSuitesList = (List<String>) configs.get(SSLConfigs.SSL_CIPHER_SUITES_CONFIG);
-            this.cipherSuites = (String[]) cipherSuitesList.toArray(new String[cipherSuitesList.size()]);
+            if(cipherSuitesList.size() > 0)
```
nit: space between `if` and `(`. 

Also, I would suggest to do `if (!cipherSuitesList.isEmpty())` :wink:

harshach commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -175,6 +175,7 @@ object Defaults {
   val SSLClientAuthRequested = "requested"
   val SSLClientAuthNone = "none"
   val SSLClientAuth = SSLClientAuthNone
+  val SSLCypherSuites = ""
```
can we make this consistent between Cypher vs Cipher. Since we've Cipher everywhere can we keep that.

harshach commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -804,6 +806,7 @@ case class KafkaConfig (props: java.util.Map[_, _]) extends AbstractConfig(Kafka
   val sslKeyManagerAlgorithm = getString(KafkaConfig.SSLKeyManagerAlgorithmProp)
   val sslTrustManagerAlgorithm = getString(KafkaConfig.SSLTrustManagerAlgorithmProp)
   val sslClientAuth = getString(KafkaConfig.SSLClientAuthProp)
+  val sslCypher = getList(KafkaConfig.SSLCipherSuitesProp)
```
same here.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SSLFactory.java:
```diff
@@ -60,12 +60,13 @@ public void configure(Map<String, ?> configs) throws KafkaException {
             List<String> cipherSuitesList = (List<String>) configs.get(SSLConfigs.SSL_CIPHER_SUITES_CONFIG);
-            this.cipherSuites = (String[]) cipherSuitesList.toArray(new String[cipherSuitesList.size()]);
+            if (!cipherSuitesList.isEmpty())
+                this.cipherSuites = cipherSuitesList.toArray(new String[cipherSuitesList.size()]);
```
Out of curiosity, why are we doing the `isEmpty` check here? We don't do the same check when setting the enabledProtocols. It seems to add noise for little benefit given how this is used.

benstopford replied:
It's because the enabled protocols have a kafka supplied default value. With the cipher suite I left it up to the JVM.

ijuma replied:
Oh I see, if you pass an empty collection, the JDK defaults are no longer used? The change makes sense then. We should probably be using the JDK defaults for enabled protocols by default too IMO, but that can be handled via KAFKA-2456

benstopford replied:
Yes - I agree. I don't think overriding the JVM defaults is a good idea unless there is a specific, documented intent.

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -201,7 +202,12 @@ object ConsumerPerformance {
       .ofType(classOf[java.lang.Integer])
-      .defaultsTo(1)
+      .defaultsTo(1)    
+    val propsFileOpt = parser.accepts("consumer-properties", "External consumer properties file.")
```
In `ConsoleConsumer` we use `consumer.config`, should we be using the same here for consistency?

ijuma commented on core/src/main/scala/kafka/tools/ProducerPerformance.scala:
```diff
@@ -174,6 +178,7 @@ object ProducerPerformance extends Logging {
     }
 
     val messageSendGapMs = options.valueOf(messageSendGapMsOpt).intValue()
+
```
Intentional new line?

ijuma commented on core/src/main/scala/kafka/tools/ProducerPerformance.scala:
```diff
@@ -38,7 +38,12 @@ import org.apache.log4j.Logger
+  }
 
+  def run(args: Array[String]) {
+    println("Starting run"+args.deep.mkString)
```
Nit: there is usually a space before and after the `+` in the Kafka codebase (there are some exceptions though).

ijuma commented:
A few mostly minor comments, the rest looks good. The command-line switch option name is something I think is important to address before merging.

ijuma commented:
LGTM

ijuma commented:
@harshach, are you interested in reviewing and merging this PR since you are now a committer? /cc @gwenshap @junrao too

benstopford commented:
tests run locally

ijuma commented on core/src/main/scala/kafka/tools/PerfConfig.scala:
```diff
@@ -39,6 +39,8 @@ class PerfConfig(args: Array[String]) {
     .defaultsTo("yyyy-MM-dd HH:mm:ss:SSS")
   val showDetailedStatsOpt = parser.accepts("show-detailed-stats", "If set, stats are reported for each reporting " +
     "interval as configured by reporting-interval")
+  val showAllStatsOpt = parser.accepts("show-all-stats", "Show intermediary stats based on the " +
```
Do we really need a new option for this? I was a bit surprised that `showDetailedStats` doesn't show the final result to be honest. You probably did this for compatibility reasons, but it would be good to get a second opinion from @gwenshap or @junrao.

benstopford replied:
I know exactly what you mean fella. In fact my first reaction was to just change the behaviour, but it's an external interface so people may have coupled to this behaviour. That's why I went down the additional option route.

ijuma replied:
Yes, I understand. It seems like we don't offer strong compatibility guarantees for these scripts though, see discussion in #242. Do you think that people would actually rely on the fact that there is no additional entry with the overall results at the end?

benstopford replied:
Quite likely not I expect, but we should overhaul these scripts if we're going down that route. I should add that the way we use jopt a little problematic. We end up overriding the system's default values just because jopt leads you down the line of providing a default value. Lots of the values used don't make much sense. For settings that map directly to kafka properties it would be better if we enforced that mapping explicitly.

junrao replied:
I agree with @ijuma . It seems it's simpler to just always print the final result instead of introducing a new option. This is a change of behavior, but it probably matches most people's expectation. ProducerPerformance is already like that.

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -234,10 +240,13 @@ object ConsumerPerformance {
     val hideHeader = options.has(hideHeaderOpt)
+    var reportingInterval = options.valueOf(reportingIntervalOpt).longValue
+    if(reportingInterval == -1)
+      reportingInterval = numMessages/10
```
To avoid the `var`, this can be done as:

```
val reportingInterval = {
  val interval = options.valueOf(reportingIntervalOpt).longValue
  if (interval == -1) numMessages / 10
  else interval
}
```

It's a bit more code (it could be done better with a utility method for this kind of thing, which is quite common), but it means that you don't have to do a find to see if this is mutated elsewhere.

benstopford replied:
yes that's much better - thanks

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -90,8 +87,12 @@ object ConsumerPerformance {
     val elapsedSecs = (endMs - startMs) / 1000.0
-    if (!config.showDetailedStats) {
-      val totalMBRead = (totalBytesRead.get * 1.0) / (1024 * 1024)
+    val totalMBRead = (totalBytesRead.get * 1.0) / (1024 * 1024)
```
Was it intentional to increase the scope of this val? It doesn't seem like it's used elsewhere.

junrao commented on core/src/main/scala/kafka/tools/PerfConfig.scala:
```diff
@@ -30,7 +30,7 @@ class PerfConfig(args: Array[String]) {
     .describedAs("size")
     .ofType(classOf[java.lang.Integer])
-    .defaultsTo(5000)
+    .defaultsTo(-1)
```
Could we add a description of the default reporting internal?

benstopford replied:
this option is a little strange as the default is overridden dynamically. I've updated the describedAs and added a comment for clarity.

junrao commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -50,19 +51,15 @@ object ConsumerPerformance {
       consumer.subscribe(List(config.topic))
       startMs = System.currentTimeMillis
-      consume(consumer, List(config.topic), config.numMessages, 1000, config, totalMessagesRead, totalBytesRead)
+      consume(consumer, List(config.topic), config.numMessages, 5000, config, totalMessagesRead, totalBytesRead)
```
Would it be better to add the timeout in the command line option?

benstopford replied:
I've parameterised this just in case.

benstopford committed:
KAFKA-2431: Various console and producer performance options

- paramaterised test-timeout in the consumer so it's a command line option
- added header for Producer/Consumer properties (so it's clear what is being overridden)
- added command line option to Producer/Consumer for freeform properties of the format key=val,key=val etc
- made linger.ms a command line option for ProducerPerf. This is a key setting.
- cleanup based on comments from Jun/Ismael

benstopford commented:
Addressed comments from above. Also added a few other bits:
- The ability to pass freeform properties of the format key=val,key=val etc. Generally useful but avoids the need for a properties file for SSL 
- Added a header that prints the properties being used (so it's clear what is being overridden as various defaults are hard coded in both ProducerPerformance and ConsumerPerformance and differ from documented defaults)
- Made linger.ms a command line option for ProducerPerf. This is an important setting.

ijuma commented:
Note that this doesn't merge cleanly anymore.

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -210,7 +230,12 @@ object ConsumerPerformance {
-    val props = new Properties
+    val props = if (options.has(propsFileOpt)) Utils.loadProps(options.valueOf(propsFileOpt)) else new Properties
+
+    if(options.has(freeformPropsOpt)) {
```
Space after `if`

ijuma commented on core/src/main/scala/kafka/tools/ProducerPerformance.scala:
```diff
@@ -160,6 +215,10 @@ object ProducerPerformance extends Logging {
     else
       new Properties()
 
+    if(options.has(freeformPropsOpt)) {
```
Space after `if`

ijuma commented on core/src/main/scala/kafka/tools/SimplePropertiesParser.scala:
```diff
@@ -0,0 +1,25 @@
+    try {
+      val argProps: immutable.Map[String, String] = freeform.split(",")
+        .map(_.split("="))
+        .map({ case Array(k, v) => (k, v)}).toMap
```
`.map { case Array(k, v) => (k, v) }.toMap` is fine, you don't need both `(` and `{`. I think you should have two `case` statement, the second one being a `case _`. Then you don't need the `catch` right? Also, instead of `JavaConversions`, prefer `JavaConverters` which means you can use the `asJava` or `asScala` methods directly.

benstopford commented:
Addressed comments from @ijuma addressed. Rebased.

benstopford committed:
KAFKA-2431: added missing license files

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -201,8 +207,22 @@ object ConsumerPerformance {
+    val freeformPropsOpt = parser.accepts("consumer-properties", "Freeform consumer properties in the format key=val,key=val,key=val...")
+      .withRequiredArg
+      .describedAs("consumer properties")
+      .ofType(classOf[java.lang.String])
```
Just noticed that `ConsoleProducer` has the following:

```
    val producerPropertyOpt = parser.accepts("producer-property", "A mechanism to pass user-defined properties in the form key=value to the producer. ")
            .withRequiredArg
            .describedAs("producer_prop")
            .ofType(classOf[String])
```

And it uses `CommandLineUtils.parseKeyValueArgs(options.valuesOf(propertyOpt))`. We probably should be using the same naming and parsing code here.

junrao commented on core/src/main/scala/kafka/tools/SimplePropertiesParser.scala:
```diff
@@ -0,0 +1,44 @@
+   * @param freeform String of format "key=val,key=val,key=val"
+   * @return
+   */
+  def propsFromFreeform(freeform: String): java.util.Map[String, String] = {
```
Instead of this, we can just use .withValuesSeparatedBy(Delimiter) in the jopt parser. See AclCommand.scala.

junrao commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -201,8 +207,22 @@ object ConsumerPerformance {
+      .describedAs("consumer properties")
+      .ofType(classOf[java.lang.String])
+      .defaultsTo("")
+    val freeformPropsOpt = parser.accepts("consumer-properties", "Freeform consumer properties in the format key=val,key=val,key=val...")
```
Do we need both consumer-properties and consumer.config? It seems that having one of the two is enough? If we want to keep both, perhaps we should do the same in ProducerPerformance.
